### PR TITLE
[FIX] MASTER AMP and onnxruntime issue with master PT

### PR DESCRIPTION
**Pull request by felixdittrich92**
This PR:

- fix amp usage for master arch in both frameworks
- fix onnxruntime issue with MASTER PT  (instead of tril triu with transpose works fine)

Any feedback is welcome :hugs: 

Closes:
#789

### CODE REVIEW

**Comment by felixdittrich92** [comment]
all slow tests passes :)

File: doctr/models/recognition/master/pytorch.py [hunk 1]
```diff
@@ -109,10 +109,10 @@ def make_source_and_target_mask(
         # [True, True, True, ..., False, False, False] -> False is masked
         target_pad_mask = (target != self.vocab_size + 2).unsqueeze(1).unsqueeze(1)  # (N, 1, 1, max_length)
         target_length = target.size(1)
-        # sub mask filled diagonal with 1 = see 0 = masked (max_length, max_length)
-        target_sub_mask = torch.tril(
-            torch.ones((target_length, target_length), dtype=torch.bool, device=source.device), diagonal=0
-        )
+        # sub mask filled diagonal with True = see and False = masked (max_length, max_length)
+        target_sub_mask = (
+            torch.triu(torch.ones((target_length, target_length), device=source.device)) == 1
```
Comment (frgfm) [inline doctr/models/recognition/master/pytorch.py:114]: If this is compatible, those options are much cleaner (best being the first):
```python
torch.triu(torch.ones((target_length, target_length), dtype=torch.bool, device=source.device)).transpose(0, 1)
```

```python
torch.triu(torch.ones((target_length, target_length), device=source.device)).to(dtype=torch.bool).transpose(0, 1)
```

File: doctr/models/recognition/transformer/pytorch.py [hunk 2]
```diff
@@ -51,7 +51,9 @@ def scaled_dot_product_attention(
 
     scores = torch.matmul(query, key.transpose(-2, -1)) / math.sqrt(query.size(-1))
     if mask is not None:
-        scores = scores.masked_fill(mask == 0, -1e9)
+        # NOTE: to ensure the ONNX compatibility, masked_fill works only with int equal condition
+        # amp fix: -6.55e4 instead of -1e9
+        scores = scores.masked_fill(mask == 0, -6.55e4)
```
Comment (frgfm) [inline doctr/models/recognition/transformer/pytorch.py:56]: why this value?
Can't we do -5e4 instead ? If there is a specific reason, add a ref to the corresponding documentation :)

File: doctr/models/recognition/transformer/tensorflow.py [hunk 3]
```diff
@@ -63,7 +66,8 @@ def scaled_dot_product_attention(
     scores = tf.matmul(query, tf.transpose(key, perm=[0, 1, 3, 2])) / math.sqrt(query.shape[-1])
     if mask is not None:
         # NOTE: to ensure the ONNX compatibility, tf.where works only with bool type condition
-        scores = tf.where(mask == False, -1e9, scores)  # noqa: E712
+        # amp fix: -6.55e4 instead of -1e9
+        scores = tf.where(mask == False, -6.55e4, scores)  # noqa: E712
```
Comment (frgfm) [inline doctr/models/recognition/transformer/tensorflow.py:70]: same

**Review by frgfm — COMMENTED** [review]
Thanks! Only a few comments

File: doctr/models/recognition/transformer/pytorch.py (hunk 2 above)
Comment (felixdittrich92) [reply doctr/models/recognition/transformer/pytorch.py:56]: changed to float('-inf') should be clear now without a comment :) -inf is the masked -> softmax = 0.0 and it is compatible with half

Comment (felixdittrich92) [reply doctr/models/recognition/transformer/pytorch.py:56]: -6.55e4 is the first value which results in 0 and can be casted internal to half

File: doctr/models/recognition/master/pytorch.py (hunk 1 above)
Comment (felixdittrich92) [reply doctr/models/recognition/master/pytorch.py:114]: Unfortunately triu works currently only with float in onnxruntime 1.11.1 opset 14 i have added a comment

Comment (felixdittrich92) [reply doctr/models/recognition/master/pytorch.py:114]: but your .to example works fine with tril nice :)

**Comment by felixdittrich92** [comment]
slow tests passes / amp tested also again
